R/tooltip.R

```diff
@@ -0,0 +1,117 @@
+#' Add a tooltip to a UI element
+#'
+#' Tooltips are useful for showing additional information when focusing (or
+#' hovering over) a UI element.
+#'
+#' @param ... Unnamed arguments can be any valid child of an [htmltools
+#'   tag][htmltools::tags]. Named arguments become HTML attributes on returned
+#'   UI element. Attributes starting with `data-bs-*` may be supplied to further
+#'   customize [tooltip
+#'   options](https://getbootstrap.com/docs/5.3/components/tooltips/).
+#' @param body Content to display in the tooltip.
+#' @param placement The placement of the tooltip relative to the target element.
+#' @param html Whether to treat `body` as HTML. WARNING: setting this `TRUE`
+#'   when the `body` is a function of user `inputs` is dangerous and not
+#'   recommended.
+#' @param sanitize Whether to sanitize HTML (only relevant when `html = TRUE`).
+#'   This can be useful if `body` is a function of user `inputs`, but should
+#'   also be treated as HTML.
+#' @param id If provided, you can use `input$id` in your server logic to
+#'   determine whether the tooltip is currently shown/hidden.
+#'
+#' @details If multiple UI elements are provided to `...`, the last element is
+#' used as the tooltip's target. If the target should contain multiple elements,
+#' then wrap those elements in a [span()] or [div()].
```

Comment:
This part feels a little weird to me. I think the first argument to `tooltip()` should be the tag that triggers the tooltip, and we could either require the user provide a tag or we could wrap a text node into a span with the correct `tabindex` attribute.

I'm inclined towards the `...` being the contents of the tooltip, which would make it easier to use `tag` functions and would be in line with `p(...)` and others. That said, I'm also in favor of an argument for the tooltip body if we're wanting to use that as a mechanism to encourage small tooltips.

Reply:
> I think the first argument to tooltip() should be the tag that triggers the tooltip

Yea, I keep going back and forth on this. I landed on the current approach basically because I think this sort of use case will be so common:

```
selectInput("foo", tooltip("Label", bsicons::bs_icon(), body = "My tooltip"), ...)
```

Obviously it's not ideal UX that the tooltip would only be attached to the icon, but at least you get a result, and can hopefully figure out to wrap everything in a `<span>` from the documentation. Plus, once we have the ability to specify virtual elements, we could figure out the bounding box of `...` and make that the tooltip's target reference.

Reply:
So the alternatives would be these?

```r
# Use tagList()
selectInput(
  inputId = "foo", 
  label = tagList("Label", tooltip(bsicons::bs_icon(), "My tooltip")),
  ...
)
```
```r
# Use a span() outside
selectInput(
  inputId = "foo", 
  label = span("Label", tooltip(bsicons::bs_icon(), "My tooltip")),
  ...
)
```
```r
# Pipe into tooltip
selectInput(
  inputId = "foo", 
  label = span("Label", bsicons::bs_icon() |> tooltip("My tooltip")),
  ...
)
# ...or...
selectInput(
  inputId = "foo", 
  label = span("Label", bsicons::bs_icon()) |> tooltip("My tooltip"),
  ...
)
```

It's when I think about piping, which seems reasonable for tooltips, it feels like `...` should be the tooltip part, but I do see your point about making it easier to put the whole tooltip as the input label.